### ISSUE TITLE
After approx 2 Invoices.EmailInvoice, api stops responding

### ISSUE THREAD

**Issue by VladSr** [XeroAPI/Xero-Net#383]
Hi there, 

when using Invoice.EmailInvoice, after 2 calls, the api no longer responds to any further requests, and always receives a timeout exception. 

Problem can be replicated using the latest .NET 4.8 / Visual Studio 2019 / Xero api 2.2.8 , and also occurs in 4.6.1 / Visual Studio 2017 / Xero api 2.2.8. The attached file is Xero api "lister" console application, adapted to send 4 emails. 

Any number of Xero API calls before the EmailInvoice calls work fine, but after approx. 2 EmailInvoice calls, the api will always timeout on any further Xero api calls, irrespective of what the Xero api call is. 

Also, the EmailInvoice api calls don't have to be immediately one after the other, other api calls can be interspersed between them. However, after the second EmailInvoice, from that point forward, all further Xero api calls time out, until the application is restarted. Then you get another two usages, then the problem repeats itself.  

Below is a log of how the error generally plays out (as expressed in the attached Xero "Lister" demo console application adaptation): 
[Program.cs.txt](https://github.com/XeroAPI/Xero-Net/files/3131049/Program.cs.txt)

Program output: 
Your organisation is called Uptime Computing Pty Ltd
There are 72 accounts
There are 2742 bank transactions
There are 640 bank transfers
There are 1 branding themes
There are 149 contacts
There are 3 credit notes
There are 1 currencies
There are 0 employees
There are 0 expense claims
There are 36 defined items
There are 254 invoices
There are 100 journal entries
There are 0 manual journal entries
There are 98 payments
There are 0 receipts
There are 6 repeating invoices
There are 11 tax rates
There are 0 tracking categories
There are 8 users
There are 10 named reports
Named:
        AgedPayablesByContact
        AgedReceivablesByContact
        BalanceSheet
        BankSummary
        BankStatement
        BudgetSummary
        ExecutiveSummary
        ProfitAndLoss
        TenNinetyNine
        TrialBalance
There are 0 published reports
29/04/2019 4:27:51 PM: emailing invoice 1
29/04/2019 4:27:53 PM: completed emailing invoice 1
29/04/2019 4:27:53 PM: emailing invoice 2
29/04/2019 4:27:54 PM: completed emailing invoice 2
29/04/2019 4:27:54 PM: emailing invoice 3
<< no response until timeout

Exception: 

System.Net.WebException
  HResult=0x80131509
  Message=The operation has timed out
  Source=System
  StackTrace:
   at System.Net.HttpWebRequest.GetResponse()
   at Xero.Api.Infrastructure.Http.HttpClient.WriteToServer(String endpoint, Byte[] data, String method, String contentType, String query)
   at Xero.Api.Infrastructure.Http.HttpClient.Post(String endpoint, Byte[] data, String contentType, String query)
   at Xero.Api.Core.Endpoints.InvoicesEndpoint.EmailInvoice(Guid invoiceId)
   at XeroNewTest.Lister.List() in C:\Users\vlad\source\repos\XeroNewTest\Program.cs:line 185
   at XeroNewTest.Program.Main(String[] args) in C:\Users\vlad\source\repos\XeroNewTest\Program.cs:line 39

Update: This appears to be the exact problem and the solution: 

http://ashwini47-tts.blogspot.com/2012/02/httpwebrequest-fails-after-few-2.html

Please let me know if you have any queries. 

Thank you, 

Vlad

**Comment by MJMortimer** [comment]
Hey @vladsr, any idea why this seems to be particularly related to the implementation of emailing invoices and doesn't happen with any of the other operations?

**Comment by VladSr** [comment]
Hey @MJMortimer<mailto:notifications@github.com>, apologies for the delayed response.

Based on the link, my guess is that a HttpWebResponse is not getting closed specifically on the path being used for Invoices.EmailInvoice().

So after the Service Point Manager uses up the default 2 connections that aren’t closed by the EmailInvoice() code path, it ceases to respond until the program is restarted, and the Service Point Manager is “reset”.

I tried to trace through and find out why other calls don’t have an issue and EmailInvoice does, but wasn’t able to find a definitive answer sorry.

In any case, I modified the code in Xero.Api.Infrastructure.Http.Response to add an additional .Close() as per below, and that seems to resolve the issue:

<< Code >>

    internal Response(HttpWebResponse inner)
    {
      StatusCode = inner.StatusCode;
      ContentLength = (int) inner.ContentLength;
      ContentType = inner.ContentType;

      var stream = inner.GetResponseStream();
      if (stream != null)
      {
        Stream = new MemoryStream();
        stream.CopyTo(Stream);
        // rewind
        Stream.Seek(0, SeekOrigin.Begin);
      }

      Headers = new Dictionary<string, string>();

      foreach (var item in inner.Headers.AllKeys)
      {
        Headers[item] = inner.Headers[item];
      }

      inner.Close(); // this is what I added
    }

<< End Code >>

With this change, I have been able to email 15 straight invoices without an issue.
